db/keyval/filedb/client.go

```diff
@@ -0,0 +1,379 @@
+
+	"github.com/pkg/errors"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/ligato/cn-infra/datasync"
+	"github.com/ligato/cn-infra/db/keyval"
+	"github.com/ligato/cn-infra/logging"
+)
+
+// Client arranges communication between file system and internal database, and is also responsible for upstream events.
+type Client struct {
+	sync.Mutex
+	log logging.Logger
+
+	// A list of filesystem paths. It may be a specific file, or the whole directory. In case a path is a directory,
+	// all files within are processed. If there are another directories inside, they are skipped.
+	cfgPaths []string
+
+	// Path where the status will be stored. Status will be stored either as json or yaml, determined by file suffix.
+	// If the field is empty or point to a directory, status is not propagated to any file.
+	statusPath string
+	// Status reader is chosen according to status file extension.
+	statusDecoder decoder.API
+
+	// Internal database mirrors changes in file system. Since the configuration can be only read, it is up to client
+	// to handle difference between configuration revisions. Every database entry consists from three values:
+	//  - path (where the configuration is written)
+	//  - data key
+	//  - data value
+	// Note: database holds all configuration, not only intended for agent with appropriate prefix
+	db database.FilesSystemDB
+
+	// Filesystem handler, provides methods to work with files/directories
+	fsHandler filesystem.API
+
+	// File system decoder API, grants access to methods needed for decoding files. Client expects
+	// the decoder to process files with specific extension
+	decoders []decoder.API
+
+	// A set of watchers for every key.
+	watchers map[string]chan keyedData
+}
+
+// NewClient initializes file watcher, database and registers paths provided via plugin configuration file
+func NewClient(cfgPaths []string, statusPath string, dcs []decoder.API, fsh filesystem.API, log logging.Logger) (*Client, error) {
+	// Init client object
+	c := &Client{
+		cfgPaths:   cfgPaths,
+		statusPath: statusPath,
+		fsHandler:  fsh,
+		watchers:   make(map[string]chan keyedData),
+		db:         database.NewDbClient(),
+		decoders:   dcs,
+		log:        log,
+	}
+
+	// Init filesystem handler
+	filePaths, err := c.fsHandler.GetFileNames(c.cfgPaths)
+	if err != nil {
+		return nil, errors.Errorf("failed to read files from provided paths: %v", err)
+	}
+	// Decode initial configuration
+	var files []*decoder.File
+	for _, filePath := range filePaths {
+		if dc := c.getFileDecoder(filePath); dc != nil {
+			byteFile, err := c.fsHandler.ReadFile(filePath)
+			if err != nil {
+				return nil, errors.Errorf("failed to read file %s content: %v", filePath, err)
+			}
+			fileEntries, err := dc.Decode(byteFile)
+			file := &decoder.File{Path: filePath, Data: fileEntries}
+			if err != nil {
+				return nil, errors.Errorf("failed to decode file %s: %v", filePath, err)
+			}
+			files = append(files, file)
```

Comment:
We could loop of file's data here, instead of saving to local `files` and do another loops below.